Should not descend into .zarr/ sub-folders

Using underlying claude produced script

<details>
<summary>create_bids_microscopy.py</summary> 

```python
#!/usr/bin/env python3
"""
Create a minimal BIDS microscopy dataset with an OME-Zarr file
Compatible with both zarr v2.x and v3.x
"""

import os
import json
import numpy as np
import zarr
import sys

# Check zarr version
zarr_version = int(zarr.__version__.split('.')[0])

SUFFIX = "SPIM"

if zarr_version >= 3:
    print(f"Detected zarr v{zarr.__version__} - using direct zarr v3 API")
    USE_OME_ZARR = False
else:
    print(f"Detected zarr v{zarr.__version__} - using ome-zarr library")
    from ome_zarr.io import parse_url
    from ome_zarr.writer import write_image
    USE_OME_ZARR = True

def create_ome_zarr_metadata_v3(group, fake_data, axes="tczyx"):
    """Create OME-Zarr metadata manually for zarr v3"""
    
    # Create multiscales metadata
    multiscales = [{
        "version": "0.4",
        "axes": [
            {"name": "t", "type": "time"},
            {"name": "c", "type": "channel"},
            {"name": "z", "type": "space", "unit": "micrometer"},
            {"name": "y", "type": "space", "unit": "micrometer"},
            {"name": "x", "type": "space", "unit": "micrometer"}
        ],
        "datasets": [
            {
                "path": "0",
                "coordinateTransformations": [
                    {
                        "type": "scale",
                        "scale": [1.0, 1.0, 1.0, 0.1, 0.1]  # t, c, z, y, x scaling
                    }
                ]
            }
        ],
        "coordinateTransformations": [
            {
                "type": "scale",
                "scale": [1.0, 1.0, 1.0, 0.1, 0.1]
            }
        ],
        "name": "example"
    }]
    
    # Create OMERO metadata
    omero = {
        "channels": [
            {
                "label": "DAPI",
                "color": "0000FF",
                "window": {"start": 0, "end": 255, "min": 0, "max": 255},
                "active": True,
                "family": "linear",
                "coefficient": 1.0
            },
            {
                "label": "GFP",
                "color": "00FF00",
                "window": {"start": 0, "end": 255, "min": 0, "max": 255},
                "active": True,
                "family": "linear",
                "coefficient": 1.0
            }
        ],
        "name": "example",
        "version": "0.4"
    }
    
    # Store metadata as group attributes
    group.attrs["multiscales"] = multiscales
    group.attrs["omero"] = omero
    
    # Create the data array
    array = group.create_array(
        "0",  # resolution level 0
        data=fake_data,
        chunks=(1, 1, 1, 64, 64)  # chunk size for each dimension
    )
    
    # Add dimension names as metadata
    array.attrs["_ARRAY_DIMENSIONS"] = list(axes)
    
    return group

def create_bids_microscopy_dataset(base_dir="bids_microscopy_example"):
    """Create a minimal BIDS microscopy dataset with fake data"""
    
    # Create directory structure
    os.makedirs(base_dir, exist_ok=True)
    
    # 1. Create dataset_description.json (required BIDS file)
    dataset_description = {
        "Name": "Example Microscopy Dataset",
        "BIDSVersion": "1.8.0",
        "DatasetType": "raw",
        "Authors": ["Example Author"],
        "License": "CC0"
    }
    
    with open(os.path.join(base_dir, "dataset_description.json"), "w") as f:
        json.dump(dataset_description, f, indent=2)
    
    # 2. Create subject directory
    subject_dir = os.path.join(base_dir, "sub-01", "micr")
    os.makedirs(subject_dir, exist_ok=True)
    
    # 3. Create fake microscopy data (5D: TCZYX)
    # Small dimensions for example: 2 timepoints, 2 channels, 3 Z-slices, 64x64 pixels
    fake_data = np.random.randint(0, 255, size=(2, 2, 3, 64, 64), dtype=np.uint8)
    
    # Add some structure to make it more realistic
    # Create a gradient pattern in the first channel
    x, y = np.meshgrid(np.linspace(0, 1, 64), np.linspace(0, 1, 64))
    gradient = ((np.sin(x * 4) + np.cos(y * 4)) * 127 + 128).astype(np.uint8)
    fake_data[:, 0, :, :, :] = gradient[np.newaxis, np.newaxis, :, :]
    
    # Create circular patterns in the second channel
    center = 32
    radius = 15
    xx, yy = np.meshgrid(np.arange(64), np.arange(64))
    circle = ((xx - center)**2 + (yy - center)**2 < radius**2) * 200
    fake_data[:, 1, :, :, :] = circle.astype(np.uint8)[np.newaxis, np.newaxis, :, :]
    
    # 4. Create OME-Zarr file
    zarr_path = os.path.join(subject_dir, f"sub-01_sample-brain_stain-DAPI_chunk-01_{SUFFIX}.ome.zarr")
    
    if USE_OME_ZARR:
        # Use ome-zarr library for zarr v2
        store = parse_url(zarr_path, mode="w").store
        root = zarr.group(store=store)
        
        # Write image with metadata
        write_image(
            image=fake_data,
            group=root,
            axes="tczyx",  # Time, Channel, Z, Y, X
            channel_names=["DAPI", "GFP"],
            omero={
                "channels": [
                    {
                        "label": "DAPI",
                        "color": "0000FF",
                        "window": {"start": 0, "end": 255}
                    },
                    {
                        "label": "GFP", 
                        "color": "00FF00",
                        "window": {"start": 0, "end": 255}
                    }
                ]
            }
        )
    else:
        # Use direct zarr v3 API
        store = zarr.storage.LocalStore(zarr_path)
        root = zarr.open_group(store=store, mode="w")
        
        # Create OME-Zarr structure with metadata
        create_ome_zarr_metadata_v3(root, fake_data)
        
        # Add .zattrs file for compatibility (some readers expect this)
        zattrs_path = os.path.join(zarr_path, ".zattrs")
        zattrs_content = {
            "multiscales": root.attrs["multiscales"],
            "omero": root.attrs["omero"]
        }
        with open(zattrs_path, "w") as f:
            json.dump(zattrs_content, f, indent=2)
    
    # 5. Create microscopy sidecar JSON file
    microscopy_json = {
        "Manufacturer": "Example Microscope Inc.",
        "ManufacturersModelName": "ExampleScope 3000",
        "InstitutionName": "Example University",
        "PixelSize": [0.1, 0.1],  # micrometers per pixel in X, Y
        "PixelSizeUnits": "µm",
        "ChunkTransformationMatrix": [[1, 0, 0, 0], [0, 1, 0, 0], [0, 0, 1, 0], [0, 0, 0, 1]],
        "ChunkTransformationMatrixAxis": ["X", "Y", "Z", "T"],
        "SampleEnvironment": "in vivo",
        "SampleExtractionProtocol": "Example protocol",
        "TissueObjective": "10x",
        "ImageAcquisitionProtocol": "Standard imaging protocol"
    }
    
    json_filename = f"sub-01_sample-brain_stain-DAPI_chunk-01_{SUFFIX}.json"
    with open(os.path.join(subject_dir, json_filename), "w") as f:
        json.dump(microscopy_json, f, indent=2)
    
    # 6. Create samples.tsv file
    samples_tsv = "sample_id\tparticipant_id\tsample_type\tspecies\n"
    samples_tsv += "brain\t01\ttissue\thomo sapiens\n"
    
    with open(os.path.join(base_dir, "samples.tsv"), "w") as f:
        f.write(samples_tsv)
    
    print(f"BIDS microscopy dataset created in: {base_dir}")
    print(f"\nFake data shape: {fake_data.shape} (TCZYX)")
    print(f"Data type: {fake_data.dtype}")
    print(f"\nZarr library: v{zarr.__version__}")
    if USE_OME_ZARR:
        print("Using ome-zarr for metadata generation")
    else:
        print("Using direct zarr v3 API for metadata generation")

def main():
    """Entry point for the script"""
    import argparse
    parser = argparse.ArgumentParser(description="Create a minimal BIDS microscopy dataset")
    parser.add_argument("--output-dir", "-o", default="bids_microscopy_example", 
                       help="Output directory for the dataset (default: bids_microscopy_example)")
    args = parser.parse_args()
    
    create_bids_microscopy_dataset(args.output_dir)

if __name__ == "__main__":
    main()
```
</details>

 I have produced  http://www.oneukrainian.com/tmp/bids_microscopy_example.20250811-1.tgz with

```shell
❯ tar -czvf bids_microscopy_example.`mdate`-1.tgz bids_microscopy_example
bids_microscopy_example/
bids_microscopy_example/dataset_description.json
bids_microscopy_example/sub-01/
bids_microscopy_example/sub-01/micr/
bids_microscopy_example/sub-01/micr/sub-01_sample-brain_stain-DAPI_chunk-01_SPIM.ome.zarr/
bids_microscopy_example/sub-01/micr/sub-01_sample-brain_stain-DAPI_chunk-01_SPIM.ome.zarr/zarr.json
bids_microscopy_example/sub-01/micr/sub-01_sample-brain_stain-DAPI_chunk-01_SPIM.ome.zarr/0/
bids_microscopy_example/sub-01/micr/sub-01_sample-brain_stain-DAPI_chunk-01_SPIM.ome.zarr/0/zarr.json
....
bids_microscopy_example/sub-01/micr/sub-01_sample-brain_stain-DAPI_chunk-01_SPIM.ome.zarr/.zattrs
bids_microscopy_example/sub-01/micr/sub-01_sample-brain_stain-DAPI_chunk-01_SPIM.json
bids_microscopy_example/samples.tsv
```
which has `.ome.zarr` which follows  new zarr version v3 to which we point from https://bids-specification.readthedocs.io/en/stable/modality-specific-files/microscopy.html#file-formats to https://ngff.openmicroscopy.org/latest/ and `zarr.json` is part within that .zarr/ . Eventually we should address 

- https://github.com/bids-standard/bids-validator/issues/226

But meanwhile: BIDS validator seems to not stop at the .zarr/ boundary and looks inside, which I think it should stop going under any .zarr/:


```shell
❯ uvx bids-validator-deno --ignoreWarnings bids_microscopy_example/
        [ERROR] SIDECAR_WITHOUT_DATAFILE A json sidecar file was found without a corresponding data file
		/sub-01/micr/sub-01_sample-brain_stain-DAPI_chunk-01_SPIM.ome.zarr/zarr.json
		/sub-01/micr/sub-01_sample-brain_stain-DAPI_chunk-01_SPIM.ome.zarr/0/zarr.json

	Please visit https://neurostars.org/search?q=SIDECAR_WITHOUT_DATAFILE for existing conversations about this issue.


          Summary:                         Available Tasks:        Available Modalities:
          4 Files, 25.4 kB                                         micr                 
          1 - Subjects 1 - Sessions                                                     

	If you have any questions, please post on https://neurostars.org/tags/bids.
```

--- a/changelog.d/20250822_091529_markiewicz_truncate_opaque_trees.md
+++ b/changelog.d/20250822_091529_markiewicz_truncate_opaque_trees.md
@@ -1,0 +1,50 @@
+<!--
+A new scriv changelog fragment.
+
+Uncomment the section that is right (remove the HTML comment wrapper).
+For top level release notes, leave all the headers commented out.
+-->
+
+<!--
+### Added
+
+- A bullet item for the Added category.
+
+-->
+<!--
+### Changed
+
+- A bullet item for the Changed category.
+
+-->
+### Fixed
+
+- Avoid descending into opaque directories (such as `*.zarr/`) when checking
+  for unused sidecars. ([#227])
+
+[#227]: https://github.com/bids-standard/bids-validator/issues/227
+
+<!--
+### Deprecated
+
+- A bullet item for the Deprecated category.
+
+-->
+<!--
+### Removed
+
+- A bullet item for the Removed category.
+
+-->
+<!--
+### Security
+
+- A bullet item for the Security category.
+
+-->
+<!--
+### Infrastructure
+
+- A bullet item for the Infrastructure category.
+
+-->

--- a/src/bids-validator.ts
+++ b/src/bids-validator.ts
@@ -6,3 +6,4 @@ const errors = result.issues.get({ severity: 'error' })
 if (errors.length) {
   Deno.exit(16)
 }
+Deno.exit(0)

--- a/src/validators/bids.ts
+++ b/src/validators/bids.ts
@@ -112,6 +112,7 @@ export async function validate(
     }
     await summary.update(context)
   }
+
   for (const check of perDSChecks) {
     await check(schema as unknown as GenericSchema, dsContext)
   }

--- a/src/validators/internal/unusedFile.ts
+++ b/src/validators/internal/unusedFile.ts
@@ -2,7 +2,7 @@ import type { GenericSchema } from '../../types/schema.ts'
 import type { BIDSFile, FileTree } from '../../types/filetree.ts'
 import type { BIDSContextDataset } from '../../schema/context.ts'
 
-function* walkFileTree(fileTree?: FileTree, dsContext: BIDSContextDataset): Generator<BIDSFile> {
+function* walkFileTree(fileTree: FileTree, dsContext: BIDSContextDataset): Generator<BIDSFile> {
   
   if (!fileTree) {
     return
@@ -26,7 +26,7 @@ export async function unusedStimulus(
   dsContext: BIDSContextDataset,
 ) {
   const stimDir = dsContext.tree.get('stimuli') as FileTree
-  const unusedStimuli = [...walkFileTree(stimDir)].filter((stimulus) => !stimulus.viewed)
+  const unusedStimuli = [...walkFileTree(stimDir, dsContext)].filter((stimulus) => !stimulus.viewed)
   if (unusedStimuli.length) {
     dsContext.issues.add({ code: 'UNUSED_STIMULUS', affects: unusedStimuli.map((s) => s.path) })
   }

--- a/src/validators/internal/unusedFile.ts
+++ b/src/validators/internal/unusedFile.ts
@@ -3,7 +3,6 @@ import type { BIDSFile, FileTree } from '../../types/filetree.ts'
 import type { BIDSContextDataset } from '../../schema/context.ts'
 
 function* walkFileTree(fileTree: FileTree, dsContext: BIDSContextDataset): Generator<BIDSFile> {
-  
   if (!fileTree) {
     return
   }
@@ -26,7 +25,9 @@ export async function unusedStimulus(
   dsContext: BIDSContextDataset,
 ) {
   const stimDir = dsContext.tree.get('stimuli') as FileTree
-  const unusedStimuli = [...walkFileTree(stimDir, dsContext)].filter((stimulus) => !stimulus.viewed)
+  const unusedStimuli = [...walkFileTree(stimDir, dsContext)].filter((stimulus) =>
+    !stimulus.viewed
+  )
   if (unusedStimuli.length) {
     dsContext.issues.add({ code: 'UNUSED_STIMULUS', affects: unusedStimuli.map((s) => s.path) })
   }
@@ -40,7 +41,7 @@ export async function sidecarWithoutDatafile(
 ) {
   const unusedSidecars = [...walkFileTree(dsContext.tree, dsContext)].filter(
     (file) => (!file.viewed && file.name.endsWith('.json') &&
-      !standalone_json.includes(file.name)), 
+      !standalone_json.includes(file.name)),
   )
   unusedSidecars.forEach((sidecar) => {
     dsContext.issues.add({ code: 'SIDECAR_WITHOUT_DATAFILE', location: sidecar.path })

--- a/src/validators/internal/unusedFile.ts
+++ b/src/validators/internal/unusedFile.ts
@@ -2,18 +2,21 @@ import type { GenericSchema } from '../../types/schema.ts'
 import type { BIDSFile, FileTree } from '../../types/filetree.ts'
 import type { BIDSContextDataset } from '../../schema/context.ts'
 
-function* walkFileTree(fileTree?: FileTree): Generator<BIDSFile> {
+function* walkFileTree(fileTree?: FileTree, dsContext: BIDSContextDataset): Generator<BIDSFile> {
+  
   if (!fileTree) {
     return
   }
+
   for (const file of fileTree.files) {
     if (!file.ignored) {
       yield file
     }
   }
+
   for (const dir of fileTree.directories) {
-    if (!dir.ignored) {
-      yield* walkFileTree(dir)
+    if (!dir.ignored && !dsContext.isPseudoFile(dir) && !dsContext.isOpaqueDirectory(dir)) {
+      yield* walkFileTree(dir, dsContext)
     }
   }
 }
@@ -35,9 +38,9 @@ export async function sidecarWithoutDatafile(
   schema: GenericSchema,
   dsContext: BIDSContextDataset,
 ) {
-  const unusedSidecars = [...walkFileTree(dsContext.tree)].filter(
+  const unusedSidecars = [...walkFileTree(dsContext.tree, dsContext)].filter(
     (file) => (!file.viewed && file.name.endsWith('.json') &&
-      !standalone_json.includes(file.name)),
+      !standalone_json.includes(file.name)), 
   )
   unusedSidecars.forEach((sidecar) => {
     dsContext.issues.add({ code: 'SIDECAR_WITHOUT_DATAFILE', location: sidecar.path })

